_posts/pl/2024-04-19-czy-wiesz-jak-uzywac-operatorow-w-javascripcie-i-typescripcie.md

```diff
@@ -0,0 +1,90 @@
+---
+layout:    post
+title:     Czy wiesz, jak używać operatorów "?" i "!" w JavaScripcie i TypeScripcie?
+date:      2024-04-19T08:00:00+01:00
+published: true
+didyouknow: true
+lang: pl
+author: pgrobelny
+image: /assets/img/posts/2024-04-19-czy-wiesz-jak-uzywac-operatorow-w-javascripcie-i-typescripcie/question-marks.webp
+tags:
+- javascript
+- typescript
+---
+
+Wykrzykniki i znaki zapytania mogą czasami wprawić w zakłopotanie, ponieważ dalece im w kodzie od swojego interpunkcyjnego znaczenia.
+
+Na początek spójrzmy na wykrzyknik. W TypeScripcie może być on używany jako tzw. non-null assertion operator. Oznacza to, że znak `!` możemy potraktować jako informację dla kompilatora, że dana zmienna nie będzie miała wartości `null` ani `undefined` w momencie użycia. 
+```typescript
+let dlugosc: number = zmienna.length;  // Błąd kompilacji, ponieważ zmienna może być null
+let dlugosc: number = zmienna!.length; // Kompilator taką linijkę kodu przepuści
+```
+Jest to do pewnego stopnia obejście mechanizmu sprawdzania typów i powinno się go używać z rozwagą. Kod taki może być mniej czytelny, a kompilator przestaje kontrolować miejsca, które mogą spowodować wystąpienie błędu w czasie działania programu. Rozważniejszym podejściem jest weryfikowanie, czy zmienna nie ma wartości `null` lub `undefined` i osobno obsługiwanie takich sytuacji.
+
+Dodatkowo, już w czystym Javascript, postawienie `!` przed zmienną sprawia, że działa on jako operator negacji logicznej. Kiedy używamy go przed wyrażeniem lub zmienną, neguje on wartość logiczną tego wyrażenia. Na przykład:
+```javascript
+let prawda = true;
+let falsz = !prawda; // falsz = false
+```
+
+Operator neguje wartość naszej zmiennej, dodatkowo konwertując ją na zmienną typu boolean. Wiedząc to, możemy używać podwójnego zaprzeczenia `!!` jako sposób konwersji wartości na korespondującą z nią wartość boolean. Poniżej kilka przykładów:
```

Comment:
Operator neguje wartość zmiennej, dodatkowo konwertując ją na zmienną typu boolean. Wiedząc to, możemy używać podwójnego zaprzeczenia `!!`, aby konwertować wartości na korespondującą z nimi wartości boolean.